packages/flutter_tools/lib/src/base/dds.dart

```diff
@@ -12,8 +12,9 @@ import 'context.dart';
 import 'io.dart' as io;
 import 'logger.dart';
 
+// TODO(fujino): This should be direct injected, rather than mutable global state.
 @visibleForTesting
-Future<dds.DartDevelopmentService> Function(
+Future<dds.DartDevelopmentService?> Function(
```

Comment:
Seems like a bug if this is ever null? In the tests you already have a `DartDevelopmentService` instantiated, can you just return that, or make a fake one?

Reply:
You're right; it's not easy to fix the tests however:

1. `DartDevelopmentService` here is a thin wrapper that lazily loads its own `dds.DartDevelopmentService` into a nullable field `_ddsInstance` by calling `ddsLauncherCallback`.
2. If in the tests, I have `ddsLauncherCallback` return a `FakeDartDevelopmentService extends Fake implements dds.DartDevelopmentService`, it his an `UnimplementedError` on this line `_ddsInstance?.done.whenComplete( ... )` https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/dds.dart#L63

Reply:
ok, implemented it and the tests pass now